# Jeux impartiaux
Dans la suite, $G=(S,A)$ est un graphe orienté acyclique.  
Si $s\in S$, est un sommet, l'ensemble $\{t\in S\,|\,(s,t)\in A\}$ des *successeurs* de $s$ est noté succ $\!(s)$.    
On autorise $G$ à être infini, mais dans ce cas on supposera que $G$ est *localement borné*
dans le sens où pour tout sommet $s$,
l'ensemble $S_s =$  succ $\!{}^*(s)=\cup_{i\in\mathbf N}$ succ $\!{}^i(s)$ des extrémités des chemins d'origine $s$, est fini.

## Fonction de Grundy et noyau

Pour $S \text{ fini } \subset\mathbf N$ on pose $\text{mex}(S)=\text{min}(\mathbf N\setminus S)$

**Prop** il existe une unique application  $g:S\rightarrow\mathbf N$ vérifiant la condition

(1) &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;$\forall s\in S,\;g(s)=$ mex $\!(g(\!$ succ $\!(s))$

**preuve**
Vues les hypothèses faites sur $G$, on peut, pour tout sommet $s\in S$, poser $\delta(s)=$ la plus grande longueur d'un chemin d'origine $s$.  
En notant $\Delta_k=\delta^{-1}([0,k])$ pour tout $k\in\mathbf N$, on remarque que $\Delta_0$ est l'ensemble des *feuilles* 
(sommets sans successeur) de $G$  et que, si $k>0$, alors succ $\!(\Delta_k)\subset\Delta_{k-1}$.  
Ceci permet de définir par récurrence sur  $k\in\mathbf N$,  une suite de fonctions $g_k:\Delta_k\rightarrow\mathbf N$ par  
- $g_0(s)=0$ si $s\in \Delta_0$ ;
- $g_k(s)=$ mex $\!(g_{k-1}(\!$ succ $\!(s)))$ pour $s\in \Delta_k$ où $k>0$.

Comme $g_k|\Delta_{k-1}=g_{k-1}$ pour $k>0$, on peut définir $g_\infty:S\rightarrow\mathbf N$ par
$\forall k,g_\infty|\Delta_k=g_k$.   
On vérifie alors aisément que l'application $g_\infty$ vérifie (1) et que c'est la seule.   $\blacksquare$

Cette application est appelée la *fonction de Grundy* de $G$ et notée $g_G$

**Prop** Il existe une unique partie $K$ de l'ensemble des sommets $S$ vérifiant la condition

(2) &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;$\forall s\in S,\;s\in K\Leftrightarrow$ succ $\!(s)\cap K=\emptyset$ 

**preuve** Existence : poser $K=g_G^{-1}(0)$. Unicité : montrer par récurrence sur $k$ que
si $K$ vérifie (2), $K\cap\Delta_k=g_G^{-1}(0)\cap\Delta_k$. $\blacksquare$

Cet ensemble est appelé le *noyau* de $G$ et noté $K_G$. 


## Jeu associé

Il se joue à deux joueurs. On fixe un sommet $s_0\in S$ (la *position initiale*).  
Le joueur 1 choisit un successeur $s_1$ de $s_0$ dans $G$, le joueur 2 choisit un successeur $s_2$ de $s_1$, etc. jusqu'à ce qu'un joueur se trouve face à une position $s_k$ sans successeur, auquel cas il a perdu. $G$ étant acyclique et localement borné on voit qu'une partie se termine toujours.

Si $s_0\not\in K_G$, le joueur 1 a une stratégie gagnante.  
En effet, il lui suffit de toujours jouer un coup $s_{2i-1}$ appartenant à $K_G$ ; le coup $s_{2i}$ joué par le joueur 2 sera alors dans $S\setminus K_G$. Comme les sommets sans successeur sont dans $K_G$, le perdant est le joueur 2. 

### Implémentation

Soit $G$ un graphe acyclique localement borné dont les sommets sont représentés  en python par un type hashable, de sorte que l'on peut représenter des ensembles finis (type $\texttt{set}$) de sommets ainsi que des dictionnaires (type $\texttt{dict}$) dont les clés sont des sommets.  

#### Utilisation de la classe $\texttt{GAME}$

On représente $G$ par une instance *G* de la classe $\texttt{GAME}$ ci-dessous.

En supposant $G$ fini,   
si *G* $=\texttt{GAME(S=}S, \texttt{succ=}\!$ succ $\!\texttt{)}$ où $S$ est l'ensemble des sommets et succ la fonction successeur $S\rightarrow\mathcal P(S)$ ou  
si *G* $=\texttt{GAME(initial=}i, \texttt{succ=}\!$ succ $\!\texttt{)}$ où $i$ est un sommet fixé tel que succ $\!{}^*(i)=S$, alors
   
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;*G*$\texttt{.S}=S$  

Sans l'hypothèse $G$ fini,   
si *G* $=\texttt{GAME(}\texttt{succ=}\!$ succ $\!,\texttt{g}=g_G\texttt{)}$ alors 

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;*G*$\texttt{.S}=\texttt{None}$   

Dans les deux cas

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;*G*$\texttt{.succ}=$ succ  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;*G*$\texttt{.g}=g_G:S\rightarrow\mathbf N$   
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;*G*$\texttt{.inK}$ est la fonction $s(\in S)\mapsto s\in K_G(\in\{\texttt{True},\texttt{False}\})$  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;*G*$\texttt{.play(}s\texttt{)}=t$ le coup joué par un joueur connaissant $K_G$  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;*G*$\texttt{.game(}s\texttt{)}$ imprime une partie jouée par deux joueurs connaissant $K_G$

Dans le premier cas l'instance *G* est dite *pleine* car elle occupe beaucoup de mémoire :  
$S$ est entièrement représenté par un ensemble python et la fonctions $g_G$ fait appel à un dictionnaire de clés $S$.   

Dans le deuxième cas, on parle d'instance *creuse*.

Enfin, si l'on connait le noyau $K_G$ mais pas la fonction de Grundy $g_G$, on peut aussi définir

*G* $=\texttt{GAME(}\texttt{succ=}\!$ succ $\!,\texttt{inK}=inK\texttt{)}$

auquel cas $G$ est une instance creuse pour laquelle on ne dispose pas de *G*$\texttt{.g}$ qui vaut $\texttt{None}$


#### Précisions sur l'implémentation

La partie intéressante est le calcul de la fonction de Grundy $g$ dans le cas d'une instance pleine. Pour chaque sommet $s$ on applique la formule (1) mais il faut s'assurer de traiter les sommets dans un ordre $s_1,\ldots,s_n$ tel que, au moment du calcul de $g(s)$, les $g(t)$, pour $t\in$ succ $\!(s)$, ont déja été calculés. Cette condition est réalisée si $\forall i,j\;(s_i,s_j)\in A \Rightarrow i>j$ (l'ordre est dit *topologique*). D'où l'algorithme 

*grundy* $(G)$   
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;$E\leftarrow S$  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**tant que** $E\neq\emptyset$ **faire**  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;$s\leftarrow$ un sommet de $E$ sans successeur dans $E$  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;définir $g(s)$  en appliquant (1)  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;$E\leftarrow E\setminus\{s\}$

Pour l'implémentation  
- on introduit un dictionnaire $d$ tel que, pour tout $s\in E$, $d[s]$ soit égal au nombre de successeurs de $s$ qui sont dans $E$ ; 
- pour trouver rapidement un élément de $F=\{s\in E\;,\;s$ n'a pas de successeur dans  $E\}$, on introduit aussi une pile qui contient précisément les éléments de $F$. 

In [2]:
class GAME:

    def __init__(self, initial = None, S = None, succ = None, g = None, inK = None):

        self.succ = succ
  
        if inK is None:

            if initial is None:

                self.S = S
            
            else:

                def parcours(s):
                    for t in succ(s):
                        if t not in self.S:
                            self.S.add(t)
                            parcours(t)

                self.S = {initial}
                parcours(initial)
                
            if g is None: # self.S != None
                pred = {s : set() for s in self.S}
                for s in self.S:
                    for t in self.succ(s):
                        pred[t].add(s)
                self._g = dict()
                self.g = lambda s: self._g[s]
                d = dict()
                pile = []
                for s in self.S:
                    d[s] = len(self.succ(s))
                    if d[s] == 0: pile.append(s)
                while pile:
                    s = pile.pop()
                    self._g[s] = 0
                    vals = set(map(self.g, self.succ(s)))
                    while self._g[s] in vals:
                        self._g[s] += 1
                    for r in pred[s]:
                        d[r] -= 1
                        if d[r] == 0: pile.append(r)
            else:
                self.g = g

            self.inK = lambda s: self.g(s) == 0 

        else:

            self.inK = inK
            self.g = None
            
    def verifK(self):
        """
        Dans le cas d'une représentation pleine,
        renvoie True si self.inK représente bien le noyau
        """

        return all(([t for t in self.succ(s) if self.inK(t)] == []) == self.inK(s) for s in self.S)

    def play(self, s):

        from random import choice

        T = self.succ(s)
        if T: return choice(list(T if self.inK(s) else {t for t in T if self.inK(t)}))    

    def game(self,s):
        
        begin = True
        while True:
            print(f'{s} {self.g(s) if self.g is not None else  "∈ K" if self.inK(s) else "∉ K"}')
            begin = False
            s = self.play(s)
            if s is None:
                break

### Blackjack mathématique

Fixons $n>0$ et $x\geqslant0$. Le jeu se compose de $2n$ cartes numérotées de $0$ à $2n-1$. Une *position* du jeu est un ensemble de $n$ de ces cartes formant le *tas principal*, les $n$ cartes restantes formant le *tas secondaire*. A tout moment, les $2n$ cartes des deux tas sont visibles des deux joueurs.    
Au début du jeu, $n$ cartes sont choisies au hasard pour former la position initiale.  

Chaque joueur à son tour joue de la manière suivante
- si le tas principal est $\{0,1,\ldots,n-1\}$ ou si la somme des numéros des cartes du tas principal est $\leqslant x$, il a perdu ;
- sinon, il échange une carte du tas principal avec une carte de numéro inférieur du tas secondaire.

#### Remarque
Le Blackjack mathématique officiel est le cas $n=6$ et $x=21$ (page 10 de [David Joyner et Ann Casey](https://citeseerx.ist.psu.edu/pdf/d8a75b8a4a6290d44ed7c9c345591d5a681ccbdb)). 



In [3]:
def Blackjack(n = 6, x = 0):

        from itertools import combinations

        S = set(combinations(range(2 * n), n))

        def succ(state):
            if sum(state) <= x:
                return set()
            E = set()
            e = set(range(2 * n))
            s = set(state)
            for i in s:
                for j in e - s:
                    if j < i:
                        E.add(tuple(sorted(s - {i} | {j})))
            return E
        
        return GAME(S=S, succ=succ)

In [4]:
blackjack = Blackjack()

blackjack.verifK()

True

In [5]:

blackjack.game((6,7,8,9,10,11))

(6, 7, 8, 9, 10, 11) 0
(5, 6, 7, 8, 10, 11) 12
(3, 5, 6, 8, 10, 11) 0
(3, 5, 6, 7, 8, 10) 4
(1, 3, 6, 7, 8, 10) 0
(1, 3, 6, 7, 8, 9) 9
(0, 1, 6, 7, 8, 9) 0
(0, 1, 2, 6, 8, 9) 3
(0, 1, 2, 3, 8, 9) 0
(0, 1, 2, 3, 6, 8) 5
(0, 1, 2, 3, 6, 7) 0
(0, 1, 2, 3, 5, 7) 1
(0, 1, 2, 3, 4, 5) 0


### Jeu des allumettes (Fort-Boyard)

On fixe $k>0$. On dispose d'un tas de $n>0$ allumettes.   
Chaque joueur, à son tour, supprime du tas $h\in[1,k-1]$ allumettes du tas. Le joueur qui se trouve devant un tas vide a perdu.

Ce jeu est associé au graphe acyclique localement borné $B=(\mathbf N, A)$ où $(s,t)\in A\Leftrightarrow s-t\in[1,k-1]$.

On vérifie que la fonction de Grundy est donnée par $g_B(s)=s$ mod $k\in[0,k-1]$  

In [6]:
def Allumettes(k):

    def succ(s):
        return {s - h for h in range(1,k) if s >= h}
    
    def g(s):
        return s % k
    
    return GAME(succ=succ, g=g)

In [7]:
Allumettes(4).game(20)

20 0
17 1
16 0
15 3
12 0
11 3
8 0
6 2
4 0
1 1
0 0


En fait, dans l'émission Fort-Boyard, le jeu est le "qui perd gagne" du jeu ci-dessus. Celui qui prend la dernière allumette a perdu

In [8]:
def FortBoyard(k):

    def succ(s):
        return {s - h for h in range(1,k) if s > h}
    
    def g(s):
        return (s - 1) % k
        
    return GAME(succ=succ, g=g)

In [9]:
FortBoyard(4).game(20)

20 3
17 0
14 1
13 0
12 3
9 0
7 2
5 0
4 3
1 0


## Somme de graphes

### Définition
$G_i=(S_i,A_i)$, $i = 1,2$, sont deux graphes orientés.  
Leur *somme cartésienne* $G=G_1+G_2$ est définie par $S_G=S_1\times S_2$ et, pour $s=(s_1,s_2)\in S\;$, succ $\!{}_G(s)=\{s_1\}\times$ succ $\!{}_{G_2}(s_2)\;\cup$ succ $\!{}_{G_1}(s_1)\times\{s_2\}$.  

Si on identifie les éléments $((s_1,s_2),s_3)$, $(s_1,(s_2,s_3))$ et $(s_1,s_2,s_3)$, cette opération est associative  
et la somme $G_1+\dots+G_p$ de $p$ graphes $G_i=(S_i,A_i)$, $i = 1,\ldots,p$, est définie par  
$S_G=S_1\times\ldots\times S_p$ et $((s_1,\ldots,s_p),(t_1,\ldots,t_p))\in A_G \Leftrightarrow\exists i:(s_i,t_i)\in A_i$ et $\forall j\neq i,\;s_j=t_j$.

Si les $G_i$ sont acycliques localement bornés, il en est de même de leur somme $G$.

### Nim-addition

Le *ou exclusif* ou *nim-addition* ou *xor*, noté $\oplus$, est la loi
de composition interne sur $\{0,1\}$ définie par $n\oplus m=n+m$ mod $2\in\{0,1\}$.    
On prolonge $\oplus$ à $\mathbf N$ en posant $n\oplus m=\Sigma_i (n_i\oplus m_i)2^i$   
où $n=\Sigma_i n_i 2^i$ et $m=\Sigma_i m_i 2^i$ 
sont les écritures en base $2$ des deux entiers $n$ et $m$ ($n_i,m_i\in\{0,1\}$).

On vérifie aisément que $(\mathbf N,\oplus)$ est un groupe commutatif, le neutre est $0$ et l'opposé $\ominus n$ d'un entier $n$ est $n$ lui-même.

### Lien entre les deux notions

**Lemme** 
$\ \forall j,n,m\in\mathbf N,\;j<n\oplus m\Rightarrow(j\oplus m<n$ ou $j\oplus n<m)$  

**preuve** 
Si $j<n\oplus m$, il existe $\ell\in\mathbf N$ tel
que $j_{\ell}<(n\oplus m)_{\ell}=n_{\ell}\oplus m_{\ell}$
et $i>\ell\Rightarrow j_i=n_i\oplus m_i$.  
On a donc $j_{\ell}=0$ et $\{n_{\ell},m_{\ell}\}=\{0,1\}$.  
Si, par exemple, $n_{\ell}=0$ et $m_{\ell}=1$, alors 
$j_{\ell}\oplus n_{\ell}=0<1=m_{\ell}$ et
$i>\ell\Rightarrow j_i\oplus n_i=m_i$  
donc  $j\oplus n<m$. $\blacksquare$

**Th** Si $G_1,\dots,G_p$ sont $p$ graphes acycliques localement bornés, alors, avec la convention $g_{G_1}\oplus\ldots\oplus g_{G_p}(s_1,\ldots,s_p)=g_{G_1}(s_1)\oplus\ldots\oplus g_{G_p}(s_p)$, on a :

$g_{G_1+\dots+G_p}=g_{G_1}\oplus\ldots\oplus g_{G_p}$

**preuve** On peut supposer que $p=2$. Il s'agit de démontrer que la fonction $g=g_1\oplus g_2$ ($g_i=g_{G_i}$), définie sur $S=S_1\times S_2$, satisfait la propriété (1) caractéristique de la fonction de Grundy de $G=G_1+G_2$, sachant que $g_i$ satisfait cette propriété pour $G_i$.    
Soit donc $s=(s_1,s_2)\in S$ de sorte que $g(s)=g_1(s_1)\oplus g_2(s_2)$.   
- $g(s)\in\mathbf N\setminus g(\!$ succ $\!{}_G(s))$  car sinon $g(s)=g(t)$ avec $t\in$ succ $\!{}_G(s)$, par exemple, $t=(s_1,t_2)$ où $t_2\in$ succ $\!{}_2(s_2)$.    
Alors $g_1(s_1)\oplus g_2(s_2)=g_1(s_1)\oplus g_2(t_2)$, d'où $g_2(s_2)=g_2(t_2)\in g_2(\!$ succ $\!{}_2(s_2))$ ce qui est faux.
- Reste à montrer que $g(s)$ est le plus petit des éléments de $\mathbf N\setminus g(\!$ succ $\!{}_G(s))$, c'est à dire que tout $j<g(s)$ appartient à $g(\!$ succ $\!{}_G(s))$.   
Et en effet, si $j<g_1(s_1)\oplus g_2(s_2)$, le lemme précédent implique que, par exemple, $j\oplus g_1(s_1)<g_2(s_2)$, de sorte que $j\oplus g_1(s_1)\in g_2(\!$ succ $\!{}_2(s_2))$ :  
$j\oplus g_1(s_1)=g_2(t_2)$ où $t_2\in$ succ $\!{}_2(s_2)$.
Finalement, $j=g_1(s_1)\oplus g_2(t_2)=g(s_1,t_2)\in g(\!$ succ $\!{}_G(s))$. $\blacksquare$ 




In [10]:
def cartesianSum(graphs):
    """
    graphs est une liste ou un tuple de graphes Gi
    chaque Gi est une instance de GRAPHE telle que Gi.g n'est pas None
    """
    p = len(graphs)

    def succ(s):
        e = set()
        for i, G in enumerate(graphs):
            for t in G.succ(s[i]):
                e.add(tuple(t if j == i else s[j] for j in range(p)))
        return e
    
    def g(s):
        k = 0
        for i, G in enumerate(graphs):
            k ^= G.g(s[i]) # en python, l'opérateur xor est "^"
        return k
    
    return GAME(succ=succ, g=g)

#def __add__(self, G):
#    return cartesianSum([self, G])
#
#GAME.__add__ = __add__


### Les jeux de Nim et de Marienbad

Le graphe *Nim*${}_0=(\mathbf{N},A_0)$ avec $(s,t)\in A_0\Leftrightarrow s<t$, est acyclique  localement borné.

In [11]:
Nim0 = GAME(succ=lambda s: set(range(s)), g = lambda s: s)

Le *jeu de Nim* est associé au graphe *Nim*${}_p=$ *Nim*${}_0+\ldots+$ *Nim*${}_0$ ($p$ fois)

In [12]:
def Nim(p):
    return cartesianSum([Nim0 for _ in range(p)]) if p else Nim0

In [13]:
Nim(5).game((1,2,3,4,4))

(1, 2, 3, 4, 4) 0
(1, 2, 3, 4, 2) 6
(1, 2, 3, 2, 2) 0
(1, 2, 3, 2, 1) 3
(1, 2, 0, 2, 1) 0
(1, 1, 0, 2, 1) 3
(1, 1, 0, 1, 1) 0
(1, 1, 0, 0, 1) 1
(1, 0, 0, 0, 1) 0
(0, 0, 0, 0, 1) 1
(0, 0, 0, 0, 0) 0


Le *jeu de Marienbad* est le "qui perd gagne" du jeu de Nim. Les feuilles (positions perdantes) sont les $p$-uplets $(s_1,\ldots,s_p)$ où tous les $s_i$ sont nuls sauf un qui vaut $1$.  
La stratégie est la même que pour le jeu de Nim, sauf en fin de partie, quand tous les $s_i$ sont dans $\{0,1\}$ sauf un.

In [14]:
def Marienbad(p):
 
    GNim = Nim(p)

    def succ(s):
        return set() if sum(s) == 1 else GNim.succ(s)

    def inK(s):
        return not GNim.inK(s) if set(s) <= {0,1} else GNim.inK(s)
    
    return GAME(succ=succ, inK=inK)
            

In [15]:
Marienbad(5).game((1,2,3,4,4))

(1, 2, 3, 4, 4) ∈ K
(1, 2, 1, 4, 4) ∉ K
(1, 0, 1, 4, 4) ∈ K
(0, 0, 1, 4, 4) ∉ K
(0, 0, 0, 4, 4) ∈ K
(0, 0, 0, 2, 4) ∉ K
(0, 0, 0, 2, 2) ∈ K
(0, 0, 0, 0, 2) ∉ K
(0, 0, 0, 0, 1) ∈ K


### Le jeu de Grundy

Une étape de ce jeu est un ensemble de tas d'allumettes. Le joueur courant choisit un tas d'au moins trois allumettes et partage ce tas en
deux sous-tas de tailles différentes. Au départ, il n'y a qu'un seul tas donc, à l'étape $k$, il y a $k$ tas. On note $G_n$ ce jeu, $n$ étant la taille du tas initial.

On représente une étape par la liste ordonnée $t=(t_1,\ldots,t_k)$ des tailles des tas.  
La situation du jeu à une telle étape est équivalente au jeu $G_{t_1}+\ldots+G_{t_k}$ donc, pour la fonction de Grundy,

$g(t_1,\ldots,t_k)=g(t_1)\oplus\ldots\oplus g(t_k)$

Pour calculer $g(n)$, on revient à la définition de $g$ : les successeurs de $(n)$ sont les $(x,y)$ tq $x\neq y$ et $x+y=n$ et

$g(n)=\text{mex}(\{g(x)\oplus g(y)\,|\,x\neq y \text{ et } x+y=n\})$



In [16]:
from functools import cache

def _Grundy():

    def succ(t):
        s = set()
        tlist = list(t)
        for i, k in enumerate(tlist):
            for h in range(1, k // 2 + (1 if k % 2 else 0)):
                s.add(tuple(sorted(tlist[:i] + [h, k - h] + tlist[i+1:])))
        return s

    @cache
    def grundy(n):
        gSucc = set()
        for x, y in succ((n,)):
            gSucc.add(grundy(x) ^ grundy(y))
        g = 0
        while g in gSucc:
            g += 1
        return g
    
    def g(t):
        c = 0
        for k in t:
            c ^= grundy(k)
        return c
    
    return GAME(succ = succ, g = g)

Grundy = _Grundy()

In [17]:
Grundy.game((21,))

(21,) 4
(9, 12) 0
(2, 7, 12) 1
(2, 3, 4, 12) 0
(2, 3, 4, 5, 7) 3
(2, 2, 3, 3, 4, 7) 0
(2, 2, 3, 3, 3, 4, 4) 1
(1, 2, 2, 3, 3, 3, 3, 4) 0
(1, 1, 2, 2, 3, 3, 3, 3, 3) 1
(1, 1, 1, 2, 2, 2, 3, 3, 3, 3) 0
(1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3) 1
(1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 3, 3) 0
(1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 3) 1
(1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2) 0


Calcul des premiers $n$ tq $G_n$ est perdant :

In [18]:
for n in range(1500):
    if Grundy.g((n,)) == 0:
        print(n, end = ' ')

0 1 2 4 7 10 20 23 26 50 53 270 273 276 282 285 288 316 334 337 340 346 359 362 365 386 389 392 566 630 633 636 639 673 676 682 685 923 926 929 932 1222 

On conjecture que tous les $G_n$, $n>1222$, sont gagnants.

## Retournements de pièces (Coin Turning)

Les *Retournements de pièces*  sont une famille de jeux caractérisée de la manière suivante :

Une position est une suite de $n$ pièces de monnaie numérotées de $0$ à $n-1$ de gauche à droite 
que l'on peut représenter par un n-uple $t$ de $0$ et de $1$ ($0$ pour pile et $1$ pour face).

Un mouvement consiste à retourner un certain nombre, non nul, de pièces de la suite. On impose que
la pièce retournée la plus à droite  passe de face à pile, les autres pièces retournées pouvant passer indifféremment 
de face à pile ou de pile à face.

On impose aussi que l'ensemble des pièces que la règle du jeu autorise à retourner ne dépend que du numéro $j$ de la pièce 
retournée la plus à droite.

On appelle *suite de Grundy* d'un jeu $G$ de retournement de pièces la suite $h(j)=h_G(j):=g_G(t^j)$ où $t^j=(\delta_{i,j})_{i\geqslant0}$ est la position du jeu pour laquelle la seule pièce sur face est celle de numéro $j$.
  

**Th**
Pour toute position $t$ d'un jeu $G$ de retournement de pièces,
$\displaystyle
g_G(t)=\bigoplus_{t_j=1}h_G(j)$

**preuve**
On ordonne les positions par l'ordre lexicographique, poids fort à droite et on procède par induction 
sur cet ordre.

Notons $j_1<\ldots<j_k$ les indices de la position $t$ tels que $t_j=1$. On note $H_j$ le jeu $G$ dans la position $t^j=(\delta_{i,j})_i$
et $H=H_{j_1}+\ldots+H_{j_k}$ dans la position $p=(t^{j_1},\ldots,t^{j_k})$. Il s'agit de montrer que $g_G(t)=g_H(p)$.

Pour cela, on remarque d'abord qu'il y a une bijection naturelle entre $\text{succ}_G(t)$ et $\text{succ}_H(p)$ : si, dans $t$, on retourne
les pièces numérotées $i_1<\ldots<i_u=j_\ell$ (avec $t_{j_\ell}=1$), on obtient la position $t'\in\text{succ}_G(t)$ à laquelle on
associe la position $p'\in\text{succ}_H(p)$ obtenue à partir de $p$ en retournant les pièces $i_1<\ldots<i_u=j_\ell$ de $H_{j_\ell}$.

Pour fixer les idées et pour ne pas introduire une tripotée d'indices difficilement compréhensibles, traitons seulement un exemple
suffisamment représentatif.

On suppose $k=3$, $(j_1,j_2,j_3)=(1,3,4)$ ; alors

$t\;=0,1,0,1,1$ et  $p$ est constitué des trois positions suivantes  

$t^1=0,1,0,0,0$,  
$t^3=0,0,0,1,0$ et   
$t^4=0,0,0,0,1$.

Supposons aussi $(i_1,i_2,i_3)=(0,1,3)$ ; alors

$t'\;=1,0,0,0,1$ et  $p'$ est constitué des trois positions suivantes  

${t^1}'=0,1,0,0,0$,  
${t^3}'=1,1,0,0,0$ et   
${t^4}'=0,0,0,0,1$.

${t^3}'<t$, donc d'après l'hypothèse de récurrence, $g_{H_3}({t^3}')=h(0)\oplus h(1)$. On en déduit  
$g_H(p')=g_{H_1}({t^1}')\oplus g_{H_3}({t^3}')\oplus g_{H_4}({t^4}')=h(1)\oplus(h(0)\oplus h(1))\oplus h(4)=h(0)\oplus h(4)$.

D'autre part on peut aussi appliquer l'hypothèse de récurrence à $t'<t$, ce qui donne    
$g_G(t')=h(0)\oplus h(4)=g_H(p')$.

Comme le raisonnement précédent s'applique à tous les $t'\in\text{succ}_G(t)$, on a 
$\{g_G(t')\,|\,t'\in\text{succ}_G(t)\}=\{g_H(p')\,|\,p'\in\text{succ}_H(p)\}$ et donc, par définition $g_G(t)=g_H(p)$. $\blacksquare$

### Implémentation

La classe $\texttt{COIN\_TURNING\_GAME(turnable, h)}$
permet de représenter les jeux de retournement de pièces.

Arguments :

$\texttt{turnable}$ est l'application qui, à $j$ associe l'ensemble des tuples $(i_1,\ldots,i_u)$, avec $i_1<\ldots<i_u=j$
tq, si $t_j=1$, on peut, dans la position $t$, retourner les pièces $i_1,\ldots,i_u$.

$\texttt{h}$ est l'application $h$ (non requise).





In [19]:
import pyoeis3
import termcolor

def color(s, c):
    return termcolor.colored(s, c, force_color=True) 

oeis = pyoeis3.OEISClient()

class COIN_TURNING_GAME(GAME):
    
    def __init__(self, turnable = None, h = None):

        self.turnable = turnable

        def succ(t):
            n = len(t)
            e = set()
            for j in range(n):
                if t[j] == 1:
                    for T in turnable(j):
                        e.add(tuple(1 - x if k in T else x for k, x in enumerate(t)))
            return e
        
        @cache
        def hfunction(j):
            if h is None:
                t = (0,) * j + (1,)
                return GAME(initial = t, succ = succ).g(t) # peu efficace
            else:
                return h(j)

        self.h = hfunction

        def g(t):
            c = 0
            for i, x in enumerate(t):
                if x == 1:
                    c ^= self.h(i)
            return c
            
        super().__init__(succ = succ, g = g)
        
    def print_h(self, n, OEIS = True):

        H = [self.h(j) for j in range(n)]
        for j in range(n):
            print(f'{j:>3}', end = '')
        print()
        for j in range(n):
            print(f'{H[j]:>3}', end = '')
        print()
        if OEIS:
            print(oeis.lookup_by_terms(H)[0].name)
        

    def game(self,s):

        for i in range(len(s)):
            print(color(f'{self.h(i):>3}', c = 'yellow'), end = '')
        print()
        while True:
            snew = self.play(s)
            for i in range(len(s)):
                x = '  1' if s[i] == 1 else '  0'
                print(x if snew is None or s[i] == snew[i] else color(x, 'green'), end = '')
            print(color(f'{self.g(s):>3}', 'yellow'))  
            if snew is None:
                break
            else:
                s = snew


### Twins

Dans ce jeu de retournement de pièces, un mouvement consiste à retourner exactement deux pièces de la suite (celle de droite passant
de face à pile).

In [20]:
turnableTwins = lambda j: {(i, j) for i in range(j)}
Twins = COIN_TURNING_GAME(turnable = turnableTwins)
Twins.print_h(10, OEIS = False)


  0  1  2  3  4  5  6  7  8  9
  0  1  2  3  4  5  6  7  8  9


Il semble que $h(j)=j$, ce qui se vérifie facilement par récurrence.
On redéfinit donc $\texttt{Twins}$ en fournissant $h$.

In [21]:
Twins =  COIN_TURNING_GAME(
    turnable = turnableTwins,
    h = lambda j: j)

Twins.game((1,) * 10)    

  0  1  2  3  4  5  6  7  8  9
  1  1  1  1  1  1  1  1  1  1  1
  1  1  1  1  0  0  1  1  1  1  0
  1  1  1  1  1  0  1  1  0  1 12
  1  1  1  1  1  1  1  1  0  0  0
  1  1  0  1  0  1  1  1  0  0  6
  1  1  0  0  0  0  1  1  0  0  0
  1  1  0  0  0  1  1  0  0  0  2
  1  1  0  0  1  1  0  0  0  0  0
  0  0  0  0  1  1  0  0  0  0  1
  0  0  0  0  0  0  0  0  0  0  0


### Mock Turtles

Ici on peut retourner une, deux ou trois pièces (celle de droite passant de face à pile).

In [22]:
turnableMockTurtles = lambda j: {(j,)} | {(i, j) for i in range(j)} | {(k, i, j) for i in range(j) for k in range(i)}
MockTurtles = COIN_TURNING_GAME(turnable = turnableMockTurtles)
MockTurtles.print_h(11)

  0  1  2  3  4  5  6  7  8  9 10
  1  2  4  7  8 11 13 14 16 19 21
Odious numbers


Montrons qu'il s'agit de la [suite des *nombres impies*](https://oeis.org/A000069) définis comme 
les entiers dont l'écriture binaire contient un nombre impair de $1$, les autres entiers étant dits *maléfiques*.

Si l'on note $I$ et $M$ les ensembles de nombres impies et maléfiques respectivement, on remarque que  
$I\oplus I\subset M, M\oplus M\subset M$ et $I\oplus M\subset I$.  

**Prop** $\forall j,\,h(j)=I_{j}:=$ le $(j+1)^{\text e}$ nombre impie.

**preuve**
Par récurrence.     
Si la propriété est vérifiée pour les $j'<j$.   
Comme $h(j)=\text{mex}(g(T))$ où $T=\text{succ}(t^j)$,
il s'agit de prouver que $[0,I_{j}[\subset g(T)$ et que $I_{j}\not\in g(T)$.

Soit donc $x<I_{j}$.
- si $x=0$, on voit que $0\in g(T)$ en retournant seulement $j$.  
- si $x\in I$, l'hypothèse de récurrence montre que $x=I_{j'}$ où $j'<j$ : on retourne $j$ et $j'$.
- et si $x\in M\setminus\{0\}$, en écrivant $x=\Sigma_{i\in X}2^i=x'\oplus x''$ où $x'=2^{i_0}$ et $x''=\Sigma_{i\in X\setminus\{i_0\}}2^i$, $x'$ et $x''$ sont dans $I$ et $<x<I_{j}$, donc $x'=I_{j'}$ et $x''=I_{j''}$ : retourner $j,j'$ et $j''$.
  
Enfin, $I_{j}\not\in g(T)$ car 
- en retournant $j$, on obtient $0$,
- en retournant deux pièces $j$ et $j'<j$ on obtient $I_{j'}<I_{j}$,
- en retournant trois pièces $j,j'$ et $j''$ on obtient $I_{j'}\oplus I_{j''}\in M\not\ni I_{j}$.
$\blacksquare$

**Corollaire** $h(j)=$ celui des deux entiers $2j$ ou $2j+1$ qui est impie. $\blacksquare$




In [23]:
MockTurtles = COIN_TURNING_GAME(
    turnable = turnableMockTurtles,
    h = lambda j: 2 * j + bin(2 * j + 1).count('1') % 2)

MockTurtles.game((1,) * 10)

  1  2  4  7  8 11 13 14 16 19
  1  1  1  1  1  1  1  1  1  1  3
  1  1  0  0  1  1  1  1  1  1  0
  0  1  0  0  1  1  1  1  0  0  2
  0  1  1  0  0  1  1  0  0  0  0
  1  0  1  0  0  0  1  0  0  0  8
  0  0  0  0  0  0  0  0  0  0  0


En fait, on peut facilement jouer à Mock Turtles sans connaitre $g$ car

**Prop** Si, dans une position de Mock Turtles, les $1$ sont aux emplacements
$j_1,\ldots,j_p$, alors   
la position est dans le noyau ssi $p$ est pair et
$j_1\oplus\ldots\oplus j_p=0$ 

**preuve**
Si $h(j_1)\oplus\ldots\oplus h(j_p)=0$ alors $p$ est pair car la nim-somme d'un nombre impair
de nombres impies est impie et donc non nulle.  
Les $h(j_k)$ sont égaux à $2j_k+$ éventuellement $1$ donc le xor des $2j_k$ est nul et donc aussi le xor des $j_k$.   
Réciproquement, si $j_1\oplus\ldots\oplus j_p=0$ et si $p$ est pair alors  $x=h(j_1)\oplus\ldots\oplus h(j_p)=0$ ou $1$ suivant 
que le nombre de $j_k$ tq $h(j_k)$ est impair est pair ou impair ; mais $p$ étant pair, $x$ est maléfique donc $x=0$.
$\blacksquare$

### Ruler

On retourne un nombre queconque de pièces consécutives (celle de droite de face à pile).

In [24]:
turnableRuler = lambda j: {tuple(range(i, j + 1)) for i in range(j + 1)}
Ruler = COIN_TURNING_GAME(turnable = turnableRuler)
Ruler.print_h(10)

  0  1  2  3  4  5  6  7  8  9
  1  2  1  4  1  2  1  8  1  2
Highest power of 2 dividing n


**Prop**
$h(j)$ est la plus grande puissance de $2$ qui divise $j+1$.

**preuve**
Pour plus de clarté, il vaut mieux travailler avec $k(x)=h(x-1)=\text{mex}(0,k(x-1), k(x-1)\oplus k(x-2),\ldots,k(x-1)\oplus\ldots\oplus k(1))$.  

On montre par récurrence sur $n$ que  $\forall n$,
1. $k(2^n)=2^n$ et
2. pour tout $x\in[1,2^n[$, $k(x)$ est une puissance de $2$ $<$ à $2^n$ et $k(x+2^n)=k(x)$.

Les détails sont laissés en exercice (montrer 2. par récurrence sur $x$). 

On conclut en remarquant que $k'(x)=$ (la plus grande puissance de $2$ qui divise $x$) vérifie aussi 1. et 2.
$\blacksquare$

In [92]:
def hRuler(j):
    b = bin(j + 1)
    return 2 ** (len(b) - b.rfind('1') - 1)
    
Ruler = COIN_TURNING_GAME(
    turnable = turnableRuler,
    h = hRuler)

Ruler.game((0,1) * 5)

  1  2  1  4  1  2  1  8  1  2
  0  1  0  1  0  1  0  1  0  1 14
  0  0  1  0  1  0  1  0  1  0  0
  0  1  0  1  0  1  0  0  1  0  5
  0  1  1  0  0  1  0  0  1  0  0
  1  0  0  1  1  0  0  0  1  0  5
  1  0  1  0  1  0  0  0  1  0  0
  1  0  0  1  0  0  0  0  1  0  4
  1  0  0  0  0  0  0  0  1  0  0
  0  1  1  1  1  1  1  1  0  0 13
  0  0  0  0  0  0  0  0  0  0  0


### Grunt

On retourne 4 pièces symétriques $0, i, k, j$ où $\displaystyle \frac{i+k}{2}=\frac{0+j}{2}$ ($j$ de face à pile).

$h(j)=\text{mex}(\{h(i)\oplus h(k)\,|\,i\neq k \text{ et } i+k=j\})$.
On reconnaît la fonction de Grundy du jeu de Grundy.

In [95]:
def turnableGrunt(j):
    if j < 3:
        return set()
    else:
        return {(0, i, j - i, j) for i in range(1, j // 2 + 1) if 2 * i != j}
    
Grunt = COIN_TURNING_GAME(
    turnable = turnableGrunt,
    h = lambda j: Grundy.g((j,)))

Grunt.game((1,) * 10)
print()
Grunt.print_h(10)

  0  0  0  1  0  2  1  0  2  1
  1  1  1  1  1  1  1  1  1  1  1
  0  1  1  0  0  1  1  0  1  1  0
  1  1  0  0  0  1  0  0  0  1  3
  0  0  0  0  0  1  0  0  1  0  0
  1  0  1  1  0  0  0  0  1  0  3
  0  0  0  1  0  0  1  0  0  0  0
  1  0  1  1  1  0  0  0  0  0  1
  0  1  1  0  0  0  0  0  0  0  0

  0  1  2  3  4  5  6  7  8  9
  0  0  0  1  0  2  1  0  2  1
Sprague-Grundy value for Grundy's game when starting with n tokens


## Retournements de pièces bidimensionnels

Dans un *jeu de retournements de pièces bidimensionnel* $G$,
une position est un plateau de dimension $n_1\times n_2$ sur lequel sont disposées
des pièces de monnaie
que l'on représente par un $n_1$-uple de $n_2$-uples de $0$ et de $1$.

Un mouvement consiste à retourner une pièce  $(i,j)$ de face à pile (appelée la pièce *droite-basse*) 
ainsi qu'un certain nombre d'autres pièces $(i',j')$ toutes 
situées  dans le rectangle $0\leqslant i'\leqslant i, 0\leqslant j'\leqslant j$.

On impose que l'ensemble des pièces que la règle du jeu autorise à retourner ne dépend que de la position $(i,j)$ de la pièce 
droite-basse retournée.

On appelle *suite de Grundy* du jeu $G$ la suite double $h(i,j)=h_G(i,j):=g_G(t^{i,j})$ où $t^{i,j}$ est la position du jeu pour laquelle la seule pièce sur face est celle à l'emplacement $(i,j)$.
  

**Th**
Pour toute position $t$ de $G$,
$\displaystyle
g_G(t)=\bigoplus_{t_{i,j}=1}h_G(i,j)$

**preuve**
analogue au cas unidimensionnel. $\blacksquare$ 

#### Exemple : Acrostic twins

Dans ce jeu de retournements bidimensionnel, un mouvement consiste à retourner deux pièces alignées horizontalement ou verticalement
(la pièce retournée la plus à droite ou la plus basse passant de face à pile).
On a alors (exercice):

$h(i,j)=\text{mex}(\{h(i',j)\,|\,i'<i\}\cup\{h(i,j')\,|\,j'<j\})=i\oplus j$.

## Produit de jeux de retournements (Tartan game)

### Définition

Si $G_1$ et $G_2$ sont deux jeux de retournements (unidimensionnels), on définit le
jeu de retournements bidimensionnel $G=G_1\times G_2$ par la règle suivante :

Un mouvement de $G$ consiste à retourner les pièces d'un ensemble $X$ de positions
si $X=X_1\times X_2$ où, pour $i=1,2$, retourner les pièces de $X_i$ est un mouvement de $G_i$
(et, bien sûr, si la pièce à la position $(\max X_1,\max X_2)$ passe de face à pile).

### Nim-multiplication

On définit par induction une loi interne $\otimes$ sur $\mathbf N$ par
$$
x\otimes y=\text{mex}(\{(x\otimes y')\oplus(x'\otimes y)\oplus(x'\otimes y')\,|\,0\leqslant x'<x, 0\leqslant y'<y\})
$$

**Th** $(\mathbf N,\oplus,\otimes)$ est un corps.

**preuve** voir [ONAG](https://www.chasse-aux-livres.fr/prix/1568811276/on-numbers-and-games-john-h-conway?reloaded=1) 
chapitre 6 dans lequel Conway démontre un résultat plus général sur les ordinaux ; on peut lire ses démonstrations 
en supposant que les ordinaux traités sont des entiers. Conway montre aussi que, pour tout $n$, 
$[0,2^{2^n}[_\mathbf N$ est un sous corps de $\mathbf N$ $\blacksquare$

In [58]:
@cache
def nimProduct(x, y):
    l = []
    for u in range(x):
        for v in range(y):
            l.append(nimProduct(x, v) ^ nimProduct(u, y) ^ nimProduct(u, v))
    g = 0
    while g in l:
        g += 1
    return g

# Affichage de la table de nim-multiplication du corps 2^(2^2)) = 16 = {0,..,15}
def print_nimProduct(n):
    print('  ', end = '')
    for j in range(n):
        print(color(f'{j:>3}', 'yellow'), end = '')
    print()         
    for i in range(n):
        print(color(f'{i:>2}', 'yellow'), end = '')
        for j in range(n):
            print(f'{nimProduct(i, j):>3}', end = '')
        print()

print_nimProduct(16)

@cache
def nimInverse(n):
    import math
    if n == 1:
        return 1
    i = math.floor(math.log2(math.log2(n)))
    a = n // (2 ** (2 ** i))
    m = nimInverse(nimProduct(n, n ^ a))
    return nimProduct(m, n ^ a)  
    

    0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15
 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
 1  0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15
 2  0  2  3  1  8 10 11  9 12 14 15 13  4  6  7  5
 3  0  3  1  2 12 15 13 14  4  7  5  6  8 11  9 10
 4  0  4  8 12  6  2 14 10 11 15  3  7 13  9  5  1
 5  0  5 10 15  2  7  8 13  3  6  9 12  1  4 11 14
 6  0  6 11 13 14  8  5  3  7  1 12 10  9 15  2  4
 7  0  7  9 14 10 13  3  4 15  8  6  1  5  2 12 11
 8  0  8 12  4 11  3  7 15 13  5  1  9  6 14 10  2
 9  0  9 14  7 15  6  1  8  5 12 11  2 10  3  4 13
10  0 10 15  5  3  9 12  6  1 11 14  4  2  8 13  7
11  0 11 13  6  7 12 10  1  9  2  4 15 14  5  3  8
12  0 12  4  8 13  1  9  5  6 10  2 14 11  7 15  3
13  0 13  6 11  9  4 15  2 14  3  8  5  7 10  1 12
14  0 14  7  9  5 11  2 12 10  4 13  3 15  1  8  6
15  0 15  5 10  1 14  4 11  2 13  7  8  3 12  6  9


### Lien entre les deux notions

**Th**
Si $G_1$ et $G_2$ sont deux jeux de retournements unidimensionnels,
$$
\forall i,j,\,h_{G_1\times G_2}(i,j)=h_{G_1}(i)\otimes h_{G_2}(j)
$$

**preuve**
Voir l'exercice 7 p. 39 de
[Games Theory, Thomas Ferguson, Part 1](https://www.math.s.chiba-u.ac.jp/~yasuda/semi/Bachelor/game-original/comb.pdf)
corrigé p. 14 de [Solutions to Part I of Game Theory](https://www.math.s.chiba-u.ac.jp/~yasuda/semi/Bachelor/game-original/solution/sol1.pdf) $\blacksquare$


In [59]:
class PRODUCT_GAME(GAME):
    
    def __init__(self, G1, G2): # G1 et G2 sont des COIN_TURNING_GAME

        self.G1 = G1
        self.G2 = G2

        def succ(t):
            n1 = len(t)
            n2 = len(t[0])
            e = set()
            for i in range(n1):
                for j in range(n2):
                    if t[i][j] == 1:
                        for T1 in G1.turnable(i):
                            for T2 in G2.turnable(j):
                                e.add(tuple(tuple(1 - x if k in T1 and l in T2 else x for l, x in enumerate(t1)) for k, t1 in enumerate(t)))
            return e

        def hfunction(i, j):
            return nimProduct(G1.h(i), G2.h(j))

        self.h = hfunction

        def g(t):
            c = 0
            for i, t1 in enumerate(t):
                for j, x in enumerate(t1):
                    if x == 1:
                        c ^= self.h(i, j)
            return c
            
        super().__init__(succ = succ, g = g)

    def print_h(self, n1, n2):
        print('  ', end = '')
        for j in range(n2):
            print(color(f'{self.G2.h(j):>3}', c = 'yellow'), end = '')
        print()         
        for i in range(n1):
            print(color(f'{self.G1.h(i):>2}', c = 'yellow'), end = '')
            for j in range(n2):
                print(color(f'{self.h(i, j):>3}', 'cyan'), end = '')
            print()

    def game(self,s):
        n1 = len(s)
        n2 = len(s[0])
        while True:
            print('  ', end = '')
            for j in range(n2):
                print(color(f'{self.G2.h(j):>3}', c = 'yellow'), end = '')
            print()
            snew = self.play(s)
            for i in range(n1):
                print(color(f'{self.G1.h(i):>2}', c = 'yellow'), end = '')
                for j in range(n2):
                    x = '  1' if s[i][j] == 1 else '  0'
                    print(x if snew is None or s[i][j] == snew[i][j] else color(x, 'green'), end = '')
                print()
            print(' ' * (3 * n2 - 4) + color(f'g = {self.g(s)}', 'yellow'))
            if snew is None:
                break
            else:
                s = snew

### Turning Corners = Twins x Twins

In [60]:
TurningCorners = PRODUCT_GAME(Twins, Twins)

TurningCorners.print_h(3, 4)
print()
TurningCorners.game(tuple((1,) * 4 for i in range(3)))

    0  1  2  3
 0  0  0  0  0
 1  0  1  2  3
 2  0  2  3  1

    0  1  2  3
 0  1  1  1  1
 1  1  1  1  1
 2  1  1  1  1
        g = 0
    0  1  2  3
 0  1  1  0  0
 1  1  1  0  0
 2  1  1  1  1
        g = 1
    0  1  2  3
 0  1  1  0  0
 1  1  0  0  1
 2  1  0  1  0
        g = 0
    0  1  2  3
 0  1  1  0  0
 1  1  1  1  1
 2  1  1  0  0
        g = 2
    0  1  2  3
 0  1  0  0  1
 1  1  0  1  0
 2  1  1  0  0
        g = 0
    0  1  2  3
 0  0  0  1  1
 1  0  0  0  0
 2  1  1  0  0
        g = 2
    0  1  2  3
 0  1  1  1  1
 1  0  0  0  0
 2  0  0  0  0
        g = 0


### Rugs = Ruler × Ruler

In [61]:
Rugs = PRODUCT_GAME(Ruler, Ruler)

Rugs.print_h(4, 5)
print()
Rugs.game(((1,0,1,0,1),
           (0,1,0,1,0),
           (1,0,1,0,1),
           (0,1,0,1,0)))

    1  2  1  4  1
 1  1  2  1  4  1
 2  2  3  2  8  2
 1  1  2  1  4  1
 4  4  8  4  6  4

    1  2  1  4  1
 1  1  0  1  0  1
 2  0  1  0  1  0
 1  1  0  1  0  1
 4  0  1  0  1  0
           g = 5
    1  2  1  4  1
 1  0  1  0  1  1
 2  1  0  1  0  0
 1  0  1  0  1  1
 4  1  0  1  0  0
           g = 0
    1  2  1  4  1
 1  0  1  1  0  0
 2  1  0  0  1  1
 1  0  1  1  0  0
 4  1  0  1  0  0
           g = 8
    1  2  1  4  1
 1  0  1  1  0  0
 2  1  0  0  0  1
 1  0  1  1  0  0
 4  1  0  1  0  0
           g = 0
    1  2  1  4  1
 1  0  1  1  0  0
 2  1  0  0  0  1
 1  1  0  0  0  0
 4  0  1  0  0  0
           g = 10
    1  2  1  4  1
 1  0  1  1  0  0
 2  1  0  0  1  0
 1  1  0  0  0  0
 4  0  1  0  0  0
           g = 0
    1  2  1  4  1
 1  0  1  1  0  0
 2  1  0  0  1  0
 1  0  0  0  0  0
 4  0  1  0  0  0
           g = 1
    1  2  1  4  1
 1  0  1  0  0  0
 2  1  0  0  1  0
 1  0  0  0  0  0
 4  0  1  0  0  0
           g = 0
    1  2  1  4  1
 1  1  0  1  1  0
 2  0  1  1  0  

## Autres jeux

### Jeu de Wythoff

Il s'agit d'une modification du jeu Nim $\!{}_2$ dans laquelle on autorise le joueur à diminuer d'une même quantité chacun des deux entiers constituant une position.

Plus précisément le jeu est associé au graphe $W$ défini par $S_W=\mathbf N\times\mathbf N$ et   
$(p,q)\in$ succ $\!{}_W(n,m)\Leftrightarrow (p=n$ et $q<m)$ ou $(p<n$ et $q=m)$ ou $(\exists k > 0, \;(p,q)=(n-k,m-k))$

On ne connait pas d'expression simple de la fonction $g_W$.  

Calculons le noyau $K_W$ restreint à $[0,7]\times[0,7]$ en coloriant en rouge les éléments de  $K_W$ et en vert ceux de son complémentaire $K_W^{\text c}$

![images/fig1.png](images/fig1.png)

En généralisant ce procédé, on démontre aisément que $K_W=\{(a_n,b_n)\;,\;n\in\mathbf N\}\cup\{(b_n,a_n)\;,\;n\in\mathbf N\}$ où   
$b_n=a_n+n$, $a_0=0$ et, pour $n>0$,
$a_n=$ min $\!([a_{n-1}+1,+\infty[\;\setminus\;\{b_i\;,\;i=0,\ldots,n-1\})$

In [101]:
def _Wythoff():

    # K = [a_0,...,a_n]
    # N = a_n
    # i = le max des j tq b_j < N
    
    K = [0]
    N = n = 0
    i = 0
   
    def succ(s):
        n, m = s
        return {(n, q) for q in range(m)} | {(p, m) for p in range(n)} | {(n - k - 1, m - k - 1) for k in range(min(n,m))}

    def inK(s):
        nonlocal K, N, i, n
        a, b = s
        if b < a: a, b = b, a
        while n < b - a:
            n += 1
            N += 1
            while i < n - 1 and K[i+1] + i + 1 == N:
                N += 1
                i += 1
            K.append(N)
        return K[b - a] == a

    return GAME(succ=succ, inK=inK)

Wythoff = _Wythoff()

In [63]:
Wythoff.game((1000,500))

(1000, 500) ∉ K
(309, 500) ∈ K
(208, 500) ∉ K
(208, 337) ∈ K
(177, 337) ∉ K
(177, 287) ∈ K
(41, 151) ∉ K
(41, 25) ∈ K
(4, 25) ∉ K
(4, 7) ∈ K
(0, 3) ∉ K
(0, 0) ∈ K


On peut aussi montrer, voir  [la thèse de G. Nivasch](https://drive.google.com/file/d/1Dft4Z_2SpegUq_VcBFxdm7A2I1cNk-t8/view) pages 13 et suivantes, que $a_n=\lfloor n\varphi\rfloor$ et $b_n=\lfloor n\varphi^2\rfloor$ où $\varphi = (1+\sqrt 5)/2$ et $\lfloor x\rfloor$ désigne la partie entière de $x$.

D'où une autre implémentation

In [64]:
def _WythoffBis():

    phi = .5 + .5 * 5.**.5

    def inK(s):
        a, b = s
        if b < a:
            a, b = b, a
        n = b - a
        return a == int(phi * n)
    
    return GAME(succ=Wythoff.succ, inK=inK)
        
WythoffBis = _WythoffBis()


In [65]:
WythoffBis.game((1000,500))

(1000, 500) ∉ K
(309, 500) ∈ K
(290, 481) ∉ K
(290, 179) ∈ K
(283, 179) ∉ K
(272, 168) ∈ K
(86, 168) ∉ K
(86, 53) ∈ K
(19, 53) ∉ K
(19, 31) ∈ K
(19, 22) ∉ K
(4, 7) ∈ K
(2, 5) ∉ K
(2, 1) ∈ K
(0, 1) ∉ K
(0, 0) ∈ K


### Jeu d'Euclide

Une position du jeu est un couple d'entiers $>0$.   
Un mouvement consiste à ôter du plus grand des deux entiers un multiple non nul du plus petit.  
Il y a donc une unique position sans successeur, le couple $(d,d)$ où $d$ est le PGCD des deux entiers initiaux.

On peut démontrer, voir [Gabriel Nivasch 2006](https://drive.google.com/file/d/1oPAuhHwGX6wHdrJCuHU5DP0eka4Ba-Rt/view), que la fonction de Grundy est donnée par $g(p,q)=\lfloor|q/p-p/q|\rfloor$.

In [100]:
def Euclide():
    
    def succ(s):
        p, q = s
        if p < q:
            return {(p, q - k * p) for k in range(1,int(q / p) + 1) if q > k * p}
        else:
            return {(p - k * q, q) for k in range(1,int(p / q) + 1) if p > k * q}
    
    def g(s):
        p, q = s
        return int(abs(p/q - q/p))
    
    return GAME(succ=succ, g=g)

In [97]:
Euclide().game((67862,21342))

(67862, 21342) 2
(25178, 21342) 0
(3836, 21342) 5
(3836, 5998) 0
(3836, 2162) 1
(1674, 2162) 0
(1674, 488) 3
(698, 488) 0
(210, 488) 1
(210, 278) 0
(210, 68) 2
(74, 68) 0
(6, 68) 11
(6, 8) 0
(6, 2) 2
(2, 2) 0
